include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+    )
+    {
+        {
+            const uint32_t tileSz = m_pgSzxy+2u*m_tilePadding;
+            for (uint32_t i = 0u; i < _residentStorageCount; ++i)
+            {
+                const auto& params = _residentStorageParams[i];
+                const uint32_t tilesPerDim = (1u<<params.tilesPerDim_log2);
+                const uint32_t extent = tilesPerDim*tileSz;
+                assert(params.formatCount>0u);
+                const E_FORMAT fmt = params.formats[0];
+                const uint32_t layers = params.layerCount;
+                m_storage.insert({params.formatClass, createVTResidentStorage(fmt, extent, layers, tilesPerDim)});
+            }
+        }
+        {
+            auto execPerFormat = [_residentStorageCount, _residentStorageParams,this] (auto f_fmtf, auto f_fmti, auto f_fmtu)
+            {
+                for (uint32_t i = 0u; i < _residentStorageCount; ++i)
+                {
+                    const auto& params = _residentStorageParams[i];
+                    for (uint32_t j = 0u; j < params.formatCount; ++j)
+                    {
+                        const E_FORMAT fmt = params.formats[j];
+                        auto* storage = m_storage[params.formatClass].get();
+                        if (isNormalizedFormat(fmt)||isFloatingPointFormat(fmt)||isScaledFormat(fmt))
+                            f_fmtf(fmt,storage);
+                        else { //integer formats
+                            if (isSignedFormat(fmt))
+                                f_fmti(fmt,storage);
+                            else
+                                f_fmtu(fmt,storage);
+                        }
+                    }
+                }
+            };
+            {
+                uint32_t fcount = 0u, icount = 0u, ucount = 0u;
+                execPerFormat([&](auto,auto) {++fcount; }, [&](auto,auto) {++icount; }, [&](auto,auto) {++ucount; });
+                m_fsamplers.views = fcount ? core::make_refctd_dynamic_array<decltype(SamplerArray::views)>(fcount) : nullptr;
+                m_isamplers.views = icount ? core::make_refctd_dynamic_array<decltype(SamplerArray::views)>(icount) : nullptr;
+                m_usamplers.views = ucount ? core::make_refctd_dynamic_array<decltype(SamplerArray::views)>(ucount) : nullptr;
+            }
+            {
+                uint32_t fi = 0u, ii = 0u, ui = 0u;
+
+                execPerFormat(
+                    [&fi, this](E_FORMAT _fmt, IVTResidentStorage* _storage) { (*m_fsamplers.views)[fi++] = _storage->createView(_fmt); },
+                    [&ii, this](E_FORMAT _fmt, IVTResidentStorage* _storage) { (*m_isamplers.views)[ii++] = _storage->createView(_fmt); },
+                    [&ui, this](E_FORMAT _fmt, IVTResidentStorage* _storage) { (*m_usamplers.views)[ui++] = _storage->createView(_fmt); }
+                );
+            }
+        }
+    }
+
+public:
+    IVirtualTexture(
+        uint32_t _pgTabSzxy_log2 = 8u,
+        uint32_t _pgTabLayers = 256u,
```

Comment:
make the default less than 256 (more like 32), right now you're creating a crazy 64MB texture

Reply:
even a value of 8 would be better because then the page table would fit very nicely in GPU L2 caches (512kb)